### PR TITLE
[TSC] Adding the Technical Release Process Document

### LINKED ISSUE #1033: [TSC] Document Technical Releasing Process
- we would like to document the process of releasing CNTT
  - deadlines for contributions.
  - deadlines for proofreading.
  - etc.

### PR DESCRIPTION
**Pull request by prabhubalan**
@rabi-abdel  - Committing the release process document  (for the issue Fixes #1033) along with some diagrams. Please review.

### CODE REVIEW

**Comment by pgoyal01** [comment]
@prabhubalan Re: proposedmodel.png
1. If all of the work is performed on the Master (green) branch, then the "Cherry Pick" has to be from the Master to the next version of the Baldy release (with Hot Fixes) and the Tag would have changed to, say, "4.0.1 Baldy".
2. The Tag has major version, major version with significant changes and minor (hot fixes) (the 3 digits) but I think that in CNTT we do not expect releases "major version with significant changes" -- Correct?
3. Typo: The bottom right "Tag" should be "5.0.0. Hallasan"

**Comment by prabhubalan** [comment]
> @prabhubalan Re: proposedmodel.png
> 
> 1. If all of the work is performed on the Master (green) branch, then the "Cherry Pick" has to be from the Master to the next version of the Baldy release (with Hot Fixes) and the Tag would have changed to, say, "4.0.1 Baldy".
@pgoyal01 - Thanks Pankaj. Updated it now
> 2. The Tag has major version, major version with significant changes and minor (hot fixes) (the 3 digits) but I think that in CNTT we do not expect releases "major version with significant changes" -- Correct?
@pgoyal01 - May not be required for Baldy. Let me request @rabi-abdel to add more comment on this. May be required for the future release, however, we will take a collective decision based on what we experience in CNTT. This will be evolving document.
> 3. Typo: The bottom right "Tag" should be "5.0.0. Hallasan"
@pgoyal01 - Thanks for pointing it out. Corrected now.

File: doc/tech/release_notes/release_process.md [hunk 1]
```diff
@@ -1,23 +1,53 @@
...
+<a name="releasemodel"></a>
+## Release Model
 
-<a name="h1"></a>
-## H1
+As a guiding principle, all the development occurs in “Master” branch. All the contribution for the milestone (especially M3) goes into ”Master”. After M4 (Proof Reading), at one point in the release candidate (RC) (for now maintain RC0) for Baldy and based on experience we shall increase or limit what get into the final release. To make this happen, branch out from “Master”, create a delivery branch (“Baldy”,”Hallasan”, “Baraque”, etc). Developers can continue their next release branch work in “Master”. At the end of each release, artifacts are “tagged” in GitHub according to the guideline and principles defined.
 
-<a name="h2"></a>
-## H2
+To deliver a hot fix (also known as patch) into the latest release, simply apply the hot fix in the delivery branch and tag it (4.0.1-Baldy). Once the hot fix is delivered, cherry pick technique will be applied to carry over the changes into the current release. As shown in the below diagram.
```
Comment (gkunz) [inline doc/tech/release_notes/release_process.md]: as discussed on the called, I'd like to propose to add a sentence here or rephrase: the mind set should always be to fix a problem on master first - irrespective of existing release branches. So, by default a hotfix should go to master first. Only if it does not apply anymore, a "local" fix on a release branch is valid.

Comment (bfcohen) [reply doc/tech/release_notes/release_process.md]: Be aware that we are treating documentation as code, which is an imperfect analogy.

Comment (prabhubalan) [reply doc/tech/release_notes/release_process.md]: I remember in the weekly call that we discussed and concluded to deliver the hot fix (also called patch) into the latest available release (which is Baldy in this case). To perform the patch delivery, we simply perform the hot fix in the delivery branch, and tag it (4.0.1-Baldy) as per the community standard practice and guideline. I remember the same pattern is followed in ONAP as well. Once the hot fix is delivered (and everyone back to happiness), we will decide how to carry over the changes into the current release branch. Most probably cherry pick technique will apply (same shown in the figure). Only during the stabilization phase, all the fixes will go in "master" and cherry pick technique will apply for taking some of the fixes onto delivery branch. Hot fix process will be more the exception than the rule.

Comment (prabhubalan) [reply doc/tech/release_notes/release_process.md]: @gkunz - I remember in the weekly call that we discussed and concluded to deliver the hot fix (also called patch) into the latest available release (which is Baldy in this case). To perform the patch delivery, we simply perform the hot fix in the delivery branch, and tag it (4.0.1-Baldy) as per the community standard practice and guideline. I remember the same pattern is followed in ONAP as well. Once the hot fix is delivered (and everyone back to happiness), we will decide how to carry over the changes into the current release branch. Most probably cherry pick technique will apply (same shown in the figure). Only during the stabilization phase, all the fixes will go in "master" and cherry pick technique will apply for taking some of the fixes onto delivery branch. Hot fix process will be more the exception than the rule.

Comment (gkunz) [reply doc/tech/release_notes/release_process.md]: Hi @prabhubalan. I am referring to best practices established in mature upstream communities, such as OpenStack. According to the OpenStack documentation [1] on stable branches, all fixes must exist on master first "*unless the change simply does not make sense on master*" and are then backported to stable branches. Please see [2] (item 4) and [3].

I do see that we are handling documentation as code here - as pointed out by @bfcohen - so we can be a bit more relaxed about not "breaking" a stable release. However, some ways-of-working could still apply.

[1] https://docs.openstack.org/project-team-guide/stable-branches.html
[2] https://docs.openstack.org/project-team-guide/stable-branches.html#appropriate-fixes
[3] https://docs.openstack.org/project-team-guide/stable-branches.html#processes

Comment (kedmison) [reply doc/tech/release_notes/release_process.md]: I agree with @gkunz.  It makes sense to 
1. refer to established practices in a mature organization and 
1. patch master first, when possible, and cherry-pick from there.  Documentation fixes may wind up going to multiple active releases, and merge history may get confusing if there is a mesh of cherry-picking going on.
1. be pragmatic about the differences in code vs. documentation.

Comment (prabhubalan) [reply doc/tech/release_notes/release_process.md]: Hi @gkunz , @kedmison - I have updated them now. Merge / Approve if you are happy with the content.

**Comment by pgoyal01** [comment]
@prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation.  @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

**Comment by prabhubalan** [comment]
> @prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation. @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

**Comment by prabhubalan** [comment]
> > @prabhubalan Can we not use the terms "hot fixes" and "patches" which are associated more with computer code than documentation. @bfcohen had made the same point earlier. If you notice, @gkunz is careful in using the term "fixes" and that may be the course to follow.

@pgoyal01 - Thanks and corrected now.

**Comment by rabiabdel** [comment]
@prabhubalan remove the label proposed from the diagram please so we get this merged. Thanks

**Comment by prabhubalan** [comment]
> @prabhubalan remove the label proposed from the diagram please so we get this merged. Thanks

@rabiabdel - I have removed them now. Thanks!

**Comment by prabhubalan** [comment]
> @prabhubalan thank you for addressing the comments!

@gkunz -Thanks Georg

**Pull request merged**